File: codes/HellDream/97.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        if(s1 == null || s2 == null|| s3 == null) return false;
        if(s3.length()!=(s1.length()+s2.length())) return false;
        boolean[][] dp = new boolean[s1.length()+1][s2.length()+1];
        dp[0][0] = true;
        for(int i=0;i<=s1.length();i++){
            dp[i][0] = s1.substring(0,i).equals(s3.substring(0,i))? true:false;
        }
        for(int i=0;i<=s2.length();i++){
            dp[0][i] = s2.substring(0,i).equals(s3.substring(0,i))? true:false;
        }
        
        for(int i=1;i<=s1.length();i++){
            for(int j=1;j<=s2.length();j++){
                if((dp[i-1][j] && s1.charAt(i-1)==s3.charAt(i+j-1))||
                   (dp[i][j-1] && s2.charAt(j-1)==s3.charAt(i+j-1))){
                    dp[i][j] = true;
                }
            }
    
        }
        return dp[s1.length()][s2.length()];
    }
}

File: codes/Garnetwzy/97.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        int m = s1.length();
        int n = s2.length();
        if(m + n != s3.length())
            return false;
        vector<vector<bool>> dp(m+1, vector<bool>(n+1, false));
        dp[0][0] = true;
        for(int i = 1; i <= n; i++) {
            dp[0][i] = dp[0][i-1] && (s2[i-1] == s3[i-1]);
        }
        for(int i = 1; i <= m; i++) {
            dp[i][0] = dp[i-1][0] && (s1[i-1] == s3[i-1]);
        }
        for(int i = 1; i <= m; i++) {
            for(int j = 1; j <= n; j++) {
                bool ans = false;
                if(s3[i+j-1] == s1[i-1] && !ans) {
                    ans = dp[i-1][j];
                }
                if(s3[i+j-1] == s2[j-1] && !ans) {
                    ans = dp[i][j-1];
                }
                dp[i][j] = ans;
            }
        }
        return dp[m][n];
    }
};